plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/data/ExecuteInsertBatchImpl.java

```diff
@@ -157,56 +161,36 @@ StringBuilder prepareQueryForStatement(
                 attributeName = DBUtils.getObjectFullName(table.getDataSource(), attribute, DBPEvaluationContext.DML);
             }
             query.append(attributeName);
+            usedAttributes.add(i);
         }
         query.append(")\n\tVALUES "); //$NON-NLS-1$
-        StringBuilder valuesPart = new StringBuilder(64);
-        valuesPart.append("(");
-        hasKey = false;
+        
+        StringJoiner valuesPart = new StringJoiner(","); //$NON-NLS-1$
         boolean skipBindValues = CommonUtils.toBoolean(options.get(DBSDataManipulator.OPTION_SKIP_BIND_VALUES));
-        for (int i = 0; i < attributes.length; i++) {
-            DBSAttributeBase attribute = attributes[i];
-            if (DBUtils.isPseudoAttribute(attribute) || (!useMultiRowInsert && (!allNulls && DBUtils.isNullValue(attributeValues[i])))) {
-                continue;
-            }
-            if (hasKey) valuesPart.append(","); //$NON-NLS-1$
-            hasKey = true;
-
-            DBDValueHandler valueHandler = handlers[i];
-            
-            if (valueHandler instanceof DBDValueBinder) {
-                valuesPart.append(((DBDValueBinder) valueHandler).makeQueryBind(attribute, attributeValues[i]));
-            } else {
-                valuesPart.append("?"); //$NON-NLS-1$
-            }
-        }
-        valuesPart.append(")"); //$NON-NLS-1$
-        if (useMultiRowInsert) {
-            for (int i = 0; i < attributeValues.length / attributes.length; i++) {
-                if (i != 0) {
-                    query.append(",");
+        for (int i = 0; i < attributeValues.length / attributes.length; i++) {
+            StringJoiner rowValuesPart = new StringJoiner(",", "(", ")"); //$NON-NLS-1$
+            for (int j : usedAttributes) {
```

Comment:
I have one idea in mind. Here in the inner loop, we continuously unbox `Integer` references. Maybe it's worthwhile to unbox them by hand once outside those loops into an `int[]` array and then iterate over it instead of `usedAttributes` list. However, I'm not sure if it will give us a substantial gain. Could you please try and test this?

Reply:
Good suggestion @akilovich int[] is faster than unboxing once we have at least 5 attributes (tested with million rows and 100 columns 100 times). I added that logic now!